mongo/src/main/java/com/redhat/lightblue/mongo/crud/MongoSafeUpdateProtocol.java

```diff
@@ -0,0 +1,349 @@
+ * prevent unbounded array growth, we will truncate the array using
+ * the docver timestamp.
+ *
+ */
+
+
+
+public class MongoSafeUpdateProtocol {
+
+    private static final Logger LOGGER=LoggerFactory.getLogger(MongoSafeUpdateProtocol.class);
+    
+    public static final String DOCVER="docver";
+
+    public static final String DOCVER_FLD=Translator.HIDDEN_SUB_PATH.toString()+"."+DOCVER;
+    public static final String DOCVER_FLD0=Translator.HIDDEN_SUB_PATH.toString()+"."+DOCVER+".0";
+
+    private static final long TOO_OLD=10l*60l*1000l; // Any docver older than 10 minutes is to old
+    
+    private final DBCollection collection;
+    private final WriteConcern writeConcern;
+    private ObjectId docVer;
+    private BulkWriteOperation bwo;
+
+    private List<Object> idsInBatch;
+
+    public MongoSafeUpdateProtocol(DBCollection collection,WriteConcern writeConcern) {
+        this.collection=collection;
+        this.writeConcern=writeConcern;
+        reset();
+    }
+
+    /**
+     * Adds a document to the current batch. The document should
+     * contain the original docver as read from the db
+     */
+    public void addDoc(DBObject doc) {
+        cleanupOldDocVer(doc);
+        DBObject q=writeReplaceQuery(doc);
+        setDocVer(doc,docVer);
+        LOGGER.debug("replaceQuery={}",q);
+        bwo.find(q).replaceOne(doc);
+        idsInBatch.add(doc.get("_id"));
+    }
+
+    /**
+     * Returns the number of queued requests
+     */
+    public int getSize() {
+        return idsInBatch.size();
+    }
+
+    /**
+     * Commits the current batch, and prepares for the next
+     * iteration. If any update operations fail, this call will detect
+     * errors and associate them with the documents using the document
+     * index.
+     */
+    public Map<Integer,Error> commit() {
+        Map<Integer,Error> results=new HashMap<>();
+        if(!idsInBatch.isEmpty()) {
+            BulkWriteResult writeResult;
+            LOGGER.debug("attempToUpdate={}",idsInBatch.size());
+            try {
+                if(writeConcern==null) {
+                    writeResult=bwo.execute();
+                } else {
+                    writeResult=bwo.execute(writeConcern);
+                }
+                LOGGER.debug("writeResult={}",writeResult);
+                if(idsInBatch.size()==writeResult.getMatchedCount()) {
+                    LOGGER.debug("Successful update");
+                } else {
+                    LOGGER.debug("notUpdated={}",idsInBatch.size()-writeResult.getMatchedCount());
+                    findConcurrentModifications(results);
+                }
+            } catch (BulkWriteException e) {
+                List<BulkWriteError> writeErrors=e.getWriteErrors();
+                if(writeErrors!=null) {
+                    for(BulkWriteError we:writeErrors) {
+                        if (we.getCode() == 11000 || we.getCode() == 11001) {
```

Comment:
Done.

On Fri, Aug 26, 2016 at 1:16 PM, Brandon Vulaj <notifications@github.com>
wrote:

> In mongo/src/main/java/com/redhat/lightblue/mongo/crud/
> MongoSafeUpdateProtocol.java
> <https://github.com/lightblue-platform/lightblue-mongo/pull/303#discussion_r76472931>
> :
>
> > +                    writeResult=bwo.execute();
> > +                } else {
> > +                    writeResult=bwo.execute(writeConcern);
> > +                }
> > +                LOGGER.debug("writeResult={}",writeResult);
> > +                if(idsInBatch.size()==writeResult.getMatchedCount()) {
> > +                    LOGGER.debug("Successful update");
> > +                } else {
> > +                    LOGGER.debug("notUpdated={}",idsInBatch.size()-writeResult.getMatchedCount());
> > +                    findConcurrentModifications(results);
> > +                }
> > +            } catch (BulkWriteException e) {
> > +                List<BulkWriteError> writeErrors=e.getWriteErrors();
> > +                if(writeErrors!=null) {
> > +                    for(BulkWriteError we:writeErrors) {
> > +                        if (we.getCode() == 11000 || we.getCode() == 11001) {
>
> Are there constants for these somewhere? Worth making them constants if
> not?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/lightblue-platform/lightblue-mongo/pull/303/files/6e592caaaf4c79fd821914fdb9640896fa145ab0#r76472931>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADgDDVuYsMQFIcb4djK0DDiYP7s4zvP3ks5qjzuigaJpZM4JtbtZ>
> .
>